### ISSUE TITLE
Keraflow, and chances to merge into Keras

### ISSUE THREAD

**Issue by ipod825** [fchollet/keras#3697]
I've written a framework based on Keras concept:
https://github.com/ipod825/keraflow

I was suggested that it would benefit users if it is merged back into Keras since the community here is large.
So I post it here for discussions. Comment is always welcome.

**Comment by farizrahman4u** [comment]
Have you looked into making custom RNNs simpler in Keras?

**Comment by ipod825** [comment]
No, it that an issue or code comment?

**Comment by farizrahman4u** [comment]
Issue.

**Comment by ipod825** [comment]
Why would you ask if I have looked into it? And I can't find which issue you are referring to. Could you tell me the issue number?

**Comment by phdowling** [comment]
Layer embedding seems like an interesting concept. When you say "simpler tensor linkage mechanism", what do you mean by that? Would this be simple to integrate into Keras without breaking things, and what sort of advantages would be gained?

**Comment by ipod825** [comment]
By "tensor linkage mechanism", I mean how theano (tensorflow) tensors are generate by Keras (or Keraflow).
I am not very familiar with Keras mechanism. I've traced its code very long before and knew that it track layer's output shape backwards. That is, the input shape of the final layer is the output shape of the penultimate layer and so on.

Also, Keras layers' keep an eye on its input and output tensors with `inbound_nodes` and `outbound_nodes`. I am not very familiar with these things, but I do think maintaining this two lists is the reason that things are complicated in Keras, which hinders me from tracing the source code again.

Personally speaking, I do think Keraflow has simpler mechanism.
User are forced to specify input shape with `Input` layers.
Then you feed an input layer to another layer, build the layer and get the output symbolic tensor (and its shape).
Repeat the process all the way to the output layer.
Then the model can simply compile with the input and output tensors without back tracking to build the layers in the model.

The advantage of a "simpler" (if you agree) tensor linkage mechanism, it lowers the barrier for developers to modify Keras (or Keraflow), which will surely increase the speed of developing.

As for integration, I believe there will be much work to integrate Kerflow into Keras. The reason is that Keraflow actually does the same things in a different manner. Actually, you could copy and paste many Keras examples (with a little modification) and run it with Keraflow. To integrate one into the other means we need to choose one major manner to work with instead of patching code from one to the other.
